### DISCUSSION TITLE
How to use Highs for MILP problem?

### DISCUSSION THREAD

**Discussion by passion4energy** [casadi/casadi#3499] [Q&A] [answered]
Hello, 
I was wondering if there is a worked example available of how to use Highs together with the Opti class?

I translated the Knapsack problem from scipy (https://docs.scipy.org/doc/scipy/tutorial/optimize.html#tutorial-optimize-milp) to Opti:
import casadi as ca
n = len(sizes)
opti = ca.Opti("conic")
s = opti.parameter(n)
opti.set_value(s, sizes)
v = opti.parameter(n)
opti.set_value(v, values)
x = opti.variable(n)
opti.subject_to(ca.dot(s, x)<=capacity)
opti.minimize(-ca.dot(v, x))
opti.solver("highs")
opti.solve()

but it throws an error:
Function solver_qpsol (0x247f9664f60)
Input 0 (h): 
[[00, 00, 00, 00, 00, 00, 00, 00], 
 [00, 00, 00, 00, 00, 00, 00, 00], 
 [00, 00, 00, 00, 00, 00, 00, 00], 
 [00, 00, 00, 00, 00, 00, 00, 00], 
 [00, 00, 00, 00, 00, 00, 00, 00], 
 [00, 00, 00, 00, 00, 00, 00, 00], 
 [00, 00, 00, 00, 00, 00, 00, 00], 
 [00, 00, 00, 00, 00, 00, 00, 00]]
Input 1 (g): [-22, -12, -16, -10, -35, -26, -42, -53]
Input 2 (a): [[21, 11, 15, 9, 34, 25, 41, 52]]
Input 3 (lba): -inf
Input 4 (uba): 100
Input 5 (lbx): [-inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf]
Input 6 (ubx): [inf, inf, inf, inf, inf, inf, inf, inf]
Input 7 (x0): [0, 0, 0, 0, 0, 0, 0, 0]
Input 8 (lam_x0): [0, 0, 0, 0, 0, 0, 0, 0]
Input 9 (lam_a0): 0
Input 10 (q): 0x8
Input 11 (p): 0x0
Traceback (most recent call last):

  File "xxx", line 356, in compat_exec
    exec(code, globals, locals)

  File "xxx", line 49, in <module>
    opti.solve()

  File "xxx", line 47860, in solve
    return _casadi.Opti_solve(self, *args)

RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
Error in Function::call for 'solver' [MXFunction] at .../casadi/core/function.cpp:1401:
Error in Function::call for 'solver' [MXFunction] at .../casadi/core/function.cpp:330:
Error in Function::operator() for 'solver_qpsol' [HighsInterface] at .../casadi/core/function.cpp:1482:
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.

Setting the option 'error_on_fail' to False does not help. Could it be that the error is thrown because H is not positive-definite but the solver was not told to ignore this part?

Defining Highs options according to https://ergo-code.github.io/HiGHS/stable/options/definitions/ like "ho={'solver': 'ipm'} or "ho={'highs.solver': 'ipm'} to exclude the quadratic term in the problem does not help or rather the options are not accepted. 

How can the options be conveyed to the solver? 
How can the fact that matrix 'h' does not exist be conveyed to the solver?
Are Input 10 and Input 11 which belong to a second order conic constraint causing the error?

Thank you in advance for your advice and recommendations.

**Comment by jgillis** [comment, marked as answer]
There may be shortcomings to our highs interface, but your example seems to be ill-posed.
The solution is Unbounded because constraints are missing.

Also, it's not working with integer values like you may be expecting, see https://github.com/casadi/casadi/wiki/FAQ:-how-to-work-with-discrete-mixed-integer-problems%3F

**Reply by passion4energy** [reply]
Many thanks for your reply. It forced me to revisit the formulation in opti:

import casadi as ca
n = len(sizes)
opti = ca.Opti("conic")
s = opti.parameter(n)
opti.set_value(s, sizes)
v = opti.parameter(n)
opti.set_value(v, values)
x = opti.variable(n)
opti.subject_to(0<=(x<=1))
opti.subject_to(0.0<=(ca.dot(s, x)<=capacity))
opti.minimize(-ca.dot(v, x))
opti.solver("highs", {"discrete": [True for i in range(n)]})
sol = opti.solve()
print("solution ", sol.value(x))

solution  [ 1.  1. -0.  1.  1.  1. -0. -0.]

which matches the values computed by highs in scipy.

Many thanks.
